lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -801,6 +802,173 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
   revng_check(TheBinary.verify(true));
 }
 
+static UpcastablePointer<model::Type>
+buildPrototype(GeneratedCodeBasicInfo &GCBI,
+               model::Binary &Binary,
+               const FunctionSummary &Summary,
+               const model::BasicBlock &Block) {
+  using namespace model;
+  using RegisterState = model::RegisterState::Values;
+
+  auto NewType = makeType<RawFunctionType>();
+  auto &CallType = *llvm::cast<RawFunctionType>(NewType.get());
+  {
+    auto ArgumentsInserter = CallType.Arguments.batch_insert();
+    auto ReturnValuesInserter = CallType.ReturnValues.batch_insert();
+
+    bool Found = false;
+    for (const auto &[PC, CallSites] : Summary.ABIResults.CallSites) {
+      if (PC != Block.Start)
+        continue;
+
+      revng_assert(!Found);
+      Found = true;
+
+      for (const auto &[Arg, RV] :
+           zipmap_range(CallSites.ArgumentsRegisters,
+                        CallSites.ReturnValuesRegisters)) {
+        auto *CSV = Arg == nullptr ? RV->first : Arg->first;
+        RegisterState RSArg = Arg == nullptr ? RegisterState::Maybe :
+                                               Arg->second;
+        RegisterState RSRV = RV == nullptr ? RegisterState::Maybe : RV->second;
+
+        auto RegisterID = ABIRegister::fromCSVName(CSV->getName(), GCBI.arch());
+        if (RegisterID == Register::Invalid || CSV == GCBI.spReg())
+          continue;
+
+        auto *CSVType = CSV->getType()->getPointerElementType();
+        auto CSVSize = CSVType->getIntegerBitWidth() / 8;
+        NamedTypedRegister TR(RegisterID);
+        TR.Type = {
+          Binary.getPrimitiveType(PrimitiveTypeKind::Generic, CSVSize), {}
+        };
+
+        if (model::RegisterState::shouldEmit(RSArg))
+          ArgumentsInserter.insert(TR);
+
+        if (model::RegisterState::shouldEmit(RSRV))
+          ReturnValuesInserter.insert(TR);
+      }
+    }
+    revng_assert(Found);
+
+    CallType.PreservedRegisters = {};
+    CallType.FinalStackOffset = 0;
+  }
+
+  return NewType;
+}
+
+static void
+finalizeModel(GeneratedCodeBasicInfo &GCBI,
```

Comment:
Can you put some comments in the body of this function?
I'll review it later, but I already now I'll ask for this.